File: typescript/src/ancilis/producers/autogen.ts
/**
 * AutoGen / AG2 framework producer (TypeScript parity with
 * `ancilis.producers.autogen`).
 *
 * Wraps AutoGen's `ConversableAgent` hook lifecycle so each inter-agent
 * message and reply becomes an Action object. Duck-typed against AutoGen —
 * no hard import. Exposes `sendHook` / `receiveHook` factories matching the
 * AutoGen-documented hook signatures, plus `attach()` to auto-wire hooks
 * against any object that has either a `register_hook` method or a
 * `hook_lists` dict.
 */

import { createHash, randomUUID } from "node:crypto";
import type { ResolvedConfig } from "../config/index.js";
import type { Action } from "../engine/action.js";
import { Engine } from "../engine/engine.js";
import { ToolRegistry, ToolStatus } from "../engine/registry.js";
import type { ToolEntry } from "../engine/registry.js";
import type { EvaluationResult } from "../engine/result.js";
import { matchesToolList } from "../engine/tool-matching.js";
import { EvidenceStore } from "../evidence/store.js";
import { canonicalJsonStringify } from "../evidence/chain.js";
import { recordAdapterUsed } from "../telemetry/index.js";
import { ProducerType } from "./protocol.js";

const PROVIDER = "autogen";
const PRODUCER_VERSION = "0.1.0";

export type AutoGenEventKind = "send" | "receive" | "reply";

export interface AutoGenEvent {
  kind: AutoGenEventKind;
  sender: string;
  recipient: string;
  message?: unknown;
  metadata?: Record<string, unknown>;
}

export interface AutoGenObservation {
  action: Action;
  evaluation: EvaluationResult;
}

export function _agentName(agent: unknown, fallback: string): string {
  if (agent === null || agent === undefined) return fallback;
  if (typeof agent === "object") {
    const name = (agent as Record<string, unknown>)["name"];
    if (typeof name === "string" && name) return name;
  }
  return String(agent);
}

export function _serializableMessage(message: unknown): unknown {
  if (
    message === null ||
    message === undefined ||
    typeof message === "string" ||
    typeof message === "number" ||
    typeof message === "boolean"
  ) {
    return message;
  }
  if (Array.isArray(message)) {
    return message.map((m) => _serializableMessage(m));
  }
  if (typeof message === "object") {
    const o = message as Record<string, unknown>;
    for (const method of ["modelDump", "model_dump", "dict", "toDict", "to_dict"]) {
      const fn = o[method];
      if (typeof fn === "function") {
        try {
          return (fn as () => unknown).call(message);
        } catch {
          continue;
        }
      }
    }
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(o)) out[k] = _serializableMessage(v);
    return out;
  }
  return String(message);
}

type SendHook = (
  sender?: unknown,
  message?: unknown,
  recipient?: unknown,
  silent?: boolean,
) => Promise<unknown>;

type ReceiveHook = (messages?: unknown) => Promise<unknown>;

export class AutoGenActionProducer {
  protected _config: ResolvedConfig;
  protected _engine: Engine;
  protected _registry: ToolRegistry;
  protected _evidenceStore: EvidenceStore;
  private _sessionId: string = randomUUID();

  constructor(
    config: ResolvedConfig,
    engine: Engine,
    registry?: ToolRegistry,
    evidenceStore?: EvidenceStore,
  ) {
    this._config = config;
    this._engine = engine;
    this._registry = registry ?? engine.registry;
    this._evidenceStore = evidenceStore ?? new EvidenceStore(config);
    recordAdapterUsed(PROVIDER);
  }

  get config(): ResolvedConfig { return this._config; }
  get producerType(): ProducerType { return ProducerType.FRAMEWORK; }
  get producerVersion(): string { return PRODUCER_VERSION; }
  get sessionId(): string { return this._sessionId; }

  protected _toolName(event: AutoGenEvent): string {
    return `${PROVIDER}:${event.kind}:${event.sender}->${event.recipient}`;
  }

  protected _buildDcCodes(): string[] {
    const codes: string[] = [];
    for (const dcCodes of this._config.dataClassifications.values()) {
      for (const code of dcCodes) {
        if (!codes.includes(code)) codes.push(code);
      }
    }
    return codes;
  }

  translate(event: AutoGenEvent): Action {
    const payload: Record<string, unknown> = {
      provider: PROVIDER,
      kind: event.kind,
      sender: event.sender,
      recipient: event.recipient,
      message: _serializableMessage(event.message),
      metadata: event.metadata ?? {},
    };
    const paramHash = createHash("sha256")
      .update(canonicalJsonStringify(payload))
      .digest("hex");
    const toolName = this._toolName(event);
    const entry = this._registry.lookup(toolName);
    return {
      actionId: randomUUID(),
      timestamp: new Date().toISOString(),
      agentId: event.sender,
      sourceType: this.producerType,
      producerType: this.producerType,
      producerVersion: this.producerVersion,
      agentOwner: this._config.agentOwner ?? null,
      actionType: "api_request",
      tool: {
        name: toolName,
        server: PROVIDER,
        descriptionHash: entry?.descriptionHash ?? null,
      },
      parameters: { raw: payload, parameterHash: paramHash },
      context: {
        dataClassifications: this._buildDcCodes(),
        activeOverlays: [...this._config.activeOverlays.keys()],
      },
    };
  }

  computeToolHash(toolIdentifier: string): string {
    return createHash("sha256").update(toolIdentifier).digest("hex");
  }

  registerTools(registry: ToolRegistry): string[] {
    return registry.getAll().map((entry) => entry.name);
  }

  protected _ensureRegistered(event: AutoGenEvent): string {
    const name = this._toolName(event);
    if (!this._registry.lookup(name)) {
      const status = matchesToolList(name, this._config.toolsAllowed)
        ? ToolStatus.APPROVED
        : ToolStatus.OBSERVED;
      const now = new Date().toISOString();
      this._registry.register({
        name,
        descriptionHash: this.computeToolHash(name),
        status,
        approvedBy: status === ToolStatus.APPROVED ? "config" : null,
        firstSeen: now,
        statusChanged: now,
      } satisfies ToolEntry);
    }
    return name;
  }

  async observe(event: AutoGenEvent): Promise<AutoGenObservation> {
    const toolName = this._ensureRegistered(event);
    const action = this.translate(event);
    const evaluation = this._engine.evaluate(action);
    await this._evidenceStore.store(evaluation, toolName);
    return { action, evaluation };
  }

  /** Hook with AutoGen's process_message_before_send signature. Returns the message unchanged. */
  sendHook(agentName?: string): SendHook {
    const fallback = agentName ?? this._config.agentName;
    return async (sender, message, recipient, silent = false): Promise<unknown> => {
      await this.observe({
        kind: "send",
        sender: _agentName(sender, fallback),
        recipient: _agentName(recipient, "unknown"),
        message,
        metadata: { silent: Boolean(silent) },
      });
      return message;
    };
  }

  /** Hook with AutoGen's process_last_received_message signature. Returns messages unchanged. */
  receiveHook(agentName?: string): ReceiveHook {
    const fallback = agentName ?? this._config.agentName;
    return async (messages): Promise<unknown> => {
      let lastMessage: unknown = null;
      let sender = "unknown";
      if (Array.isArray(messages) && messages.length > 0) {
        lastMessage = messages[messages.length - 1];
        if (lastMessage && typeof lastMessage === "object") {
          const lm = lastMessage as Record<string, unknown>;
          const candidate = lm["name"] ?? lm["role"];
          if (candidate !== undefined) sender = String(candidate);
        }
      }
      await this.observe({
        kind: "receive",
        sender,
        recipient: fallback,
        message: lastMessage,
        metadata: {},
      });
      return messages;
    };
  }

  /**
   * Auto-wire send + receive hooks against a ConversableAgent-shaped object.
   *
   * Tries `register_hook(name, fn)` first (newer AG2), then `hook_lists` dict
   * (older autogen), then falls back to direct attribute assignment on the
   * object. Returns the dict of registered callables so callers can introspect.
   */
  attach(agent: unknown, agentName?: string): Record<string, SendHook | ReceiveHook> {
    const targetName = agentName ?? _agentName(agent, this._config.agentName);
    const send = this.sendHook(targetName);
    const receive = this.receiveHook(targetName);
    const registered: Record<string, SendHook | ReceiveHook> = {
      process_message_before_send: send,
      process_last_received_message: receive,
    };

    if (agent === null || agent === undefined || typeof agent !== "object") {
      return registered;
    }
    const a = agent as Record<string, unknown>;
    const registerHook = a["register_hook"];
    if (typeof registerHook === "function") {
      for (const [name, fn] of Object.entries(registered)) {
        try {
          (registerHook as (...args: unknown[]) => unknown).call(agent, name, fn);
        } catch {
          // Some implementations expect named args { hookable_method, hook }.
          (registerHook as (...args: unknown[]) => unknown).call(agent, {
            hookable_method: name,
            hook: fn,
          });
        }
      }
      return registered;
    }
    const hookLists = a["hook_lists"];
    if (hookLists && typeof hookLists === "object" && !Array.isArray(hookLists)) {
      const hl = hookLists as Record<string, unknown[]>;
      for (const [name, fn] of Object.entries(registered)) {
        if (!Array.isArray(hl[name])) hl[name] = [];
        (hl[name] as unknown[]).push(fn);
      }
      return registered;
    }
    // Fallback: bare attribute assignment.
    for (const [name, fn] of Object.entries(registered)) a[name] = fn;
    return registered;
  }
}

File: python/src/ancilis/producers/autogen.py
"""AutoGen / AG2 framework producer.

Wraps Microsoft AutoGen / AG2 ``ConversableAgent`` hook lifecycle so each
inter-agent message and agent reply becomes an Action object. Duck-typed
against ``autogen`` / ``ag2`` — the producer does not import either, so the
SDK is not required at runtime.

Typical wiring::

    from autogen import ConversableAgent
    from ancilis.producers import AutoGenActionProducer

    producer = AutoGenActionProducer(config=cfg, engine=engine)
    user = ConversableAgent("user", ...)
    assistant = ConversableAgent("assistant", ...)
    producer.attach(assistant)  # registers send/receive hooks
    user.initiate_chat(assistant, message="hello")

The ``attach`` method registers hooks against the AutoGen-documented hook
names: ``process_message_before_send`` and ``process_last_received_message``.
For environments where you'd rather wire the closures yourself, use
``send_hook(agent_name=...)`` and ``receive_hook(agent_name=...)`` directly.
"""

from __future__ import annotations

import hashlib
import json
import uuid
from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from ancilis.config import ResolvedConfig
from ancilis.engine.action import Action, ActionContext, ActionParameters, ToolInfo
from ancilis.engine.engine import Engine
from ancilis.engine.registry import ToolEntry, ToolRegistry, ToolStatus
from ancilis.engine.result import EvaluationResult
from ancilis.evidence.store import EvidenceStore
from ancilis.producers.protocol import ProducerType
from ancilis.telemetry import record_adapter_used

PROVIDER = "autogen"
PRODUCER_VERSION = "0.1.0"


@dataclass
class AutoGenEvent:
    """Normalized AutoGen hook event (send / receive / reply)."""

    kind: str  # "send" | "receive" | "reply"
    sender: str
    recipient: str
    message: Any = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class AutoGenObservation:
    action: Action
    evaluation: EvaluationResult


def _agent_name(agent: Any, fallback: str) -> str:
    if agent is None:
        return fallback
    name = getattr(agent, "name", None)
    if isinstance(name, str) and name:
        return name
    return str(agent)


def _serializable_message(message: Any) -> Any:
    if message is None or isinstance(message, (str, int, float, bool)):
        return message
    if isinstance(message, dict):
        return {k: _serializable_message(v) for k, v in message.items()}
    if isinstance(message, list):
        return [_serializable_message(m) for m in message]
    for attr in ("model_dump", "dict", "to_dict"):
        method = getattr(message, attr, None)
        if callable(method):
            try:
                return method()
            except Exception:  # pragma: no cover - upstream object misbehaving
                continue
    return repr(message)


class AutoGenActionProducer:
    """Underlying producer that translates ``AutoGenEvent`` objects into Actions."""

    def __init__(
        self,
        config: ResolvedConfig,
        engine: Engine,
        registry: ToolRegistry | None = None,
        evidence_store: EvidenceStore | None = None,
    ) -> None:
        self._config = config
        self._engine = engine
        self._registry = registry or engine.registry
        self._evidence_store = evidence_store if evidence_store is not None else EvidenceStore(config)
        self._session_id = str(uuid.uuid4())
        record_adapter_used(PROVIDER)

    @property
    def session_id(self) -> str:
        return self._session_id

    @property
    def producer_type(self) -> ProducerType:
        return ProducerType.FRAMEWORK

    @property
    def producer_version(self) -> str:
        return PRODUCER_VERSION

    def _tool_name(self, event: AutoGenEvent) -> str:
        return f"{PROVIDER}:{event.kind}:{event.sender}->{event.recipient}"

    def translate(self, raw_invocation: AutoGenEvent) -> Action:
        payload = {
            "provider": PROVIDER,
            "kind": raw_invocation.kind,
            "sender": raw_invocation.sender,
            "recipient": raw_invocation.recipient,
            "message": _serializable_message(raw_invocation.message),
            "metadata": raw_invocation.metadata,
        }
        param_hash = hashlib.sha256(
            json.dumps(payload, sort_keys=True, default=repr).encode()
        ).hexdigest()
        tool_name = self._tool_name(raw_invocation)
        entry = self._registry.lookup(tool_name)
        dc_codes: list[str] = []
        for codes in self._config.data_classifications.values():
            for code in codes:
                if code not in dc_codes:
                    dc_codes.append(code)
        return Action(
            action_id=str(uuid.uuid4()),
            timestamp=datetime.now(timezone.utc).isoformat(),
            agent_id=raw_invocation.sender,
            source_type=self.producer_type.value,
            agent_owner=self._config.agent_owner or None,
            action_type="api_request",
            tool=ToolInfo(
                name=tool_name,
                server=PROVIDER,
                description_hash=entry.description_hash if entry else None,
            ),
            parameters=ActionParameters(raw=payload, parameter_hash=param_hash),
            context=ActionContext(
                data_classifications=dc_codes,
                active_overlays=list(self._config.active_overlays.keys()),
            ),
            producer_type=self.producer_type.value,
            producer_version=self.producer_version,
        )

    def compute_tool_hash(self, tool_identifier: Any) -> str:
        return hashlib.sha256(str(tool_identifier).encode()).hexdigest()

    def register_tools(self, registry: ToolRegistry) -> list[str]:
        return [entry.name for entry in registry.get_all()]

    def _ensure_registered(self, event: AutoGenEvent) -> str:
        name = self._tool_name(event)
        if self._registry.lookup(name) is None:
            status = (
                ToolStatus.APPROVED
                if name in self._config.tools_allowed
                else ToolStatus.OBSERVED
            )
            self._registry.register(
                ToolEntry(
                    name=name,
                    description_hash=self.compute_tool_hash(name),
                    status=status,
                    approved_by="config" if status == ToolStatus.APPROVED else None,
                )
            )
        return name

    def observe(self, event: AutoGenEvent) -> AutoGenObservation:
        tool_name = self._ensure_registered(event)
        action = self.translate(event)
        evaluation = self._engine.evaluate(action)
        self._evidence_store.store(evaluation, tool_name=tool_name)
        return AutoGenObservation(action=action, evaluation=evaluation)

    # --- AutoGen-shaped hook factories ---

    def send_hook(
        self, agent_name: str | None = None
    ) -> Callable[..., Any]:
        """Hook with AutoGen's ``process_message_before_send`` signature.

        Signature: ``(sender, message, recipient, silent) -> message``. Returns
        the message unchanged so AutoGen forwards it as-is.
        """

        fallback = agent_name or self._config.agent_name

        def hook(sender: Any = None, message: Any = None, recipient: Any = None, silent: bool = False) -> Any:
            event = AutoGenEvent(
                kind="send",
                sender=_agent_name(sender, fallback),
                recipient=_agent_name(recipient, "unknown"),
                message=message,
                metadata={"silent": bool(silent)},
            )
            self.observe(event)
            return message

        return hook

    def receive_hook(
        self, agent_name: str | None = None
    ) -> Callable[..., Any]:
        """Hook with AutoGen's ``process_last_received_message`` signature.

        Signature: ``(messages) -> messages``. The hook treats ``messages[-1]``
        as the latest received message and emits a ``receive`` event for it.
        """

        fallback = agent_name or self._config.agent_name

        def hook(messages: Any = None) -> Any:
            last_message: Any = None
            sender = "unknown"
            if isinstance(messages, list) and messages:
                last_message = messages[-1]
                if isinstance(last_message, dict):
                    sender = str(last_message.get("name") or last_message.get("role") or "unknown")
            event = AutoGenEvent(
                kind="receive",
                sender=sender,
                recipient=fallback,
                message=last_message,
                metadata={},
            )
            self.observe(event)
            return messages

        return hook

    def attach(
        self,
        agent: Any,
        *,
        agent_name: str | None = None,
    ) -> dict[str, Callable[..., Any]]:
        """Register send + receive hooks on a ``ConversableAgent``-shaped object.

        Returns the dict of hook callables registered, keyed by AutoGen hook
        name. Works against both the older ``hook_lists`` dict and the newer
        ``register_hook`` method on AG2; falls back to direct attribute
        assignment if neither is found.
        """

        send = self.send_hook(agent_name=agent_name or _agent_name(agent, self._config.agent_name))
        receive = self.receive_hook(agent_name=agent_name or _agent_name(agent, self._config.agent_name))
        registered: dict[str, Callable[..., Any]] = {
            "process_message_before_send": send,
            "process_last_received_message": receive,
        }

        register_hook = getattr(agent, "register_hook", None)
        if callable(register_hook):
            for hook_name, fn in registered.items():
                try:
                    register_hook(hookable_method=hook_name, hook=fn)
                except TypeError:
                    register_hook(hook_name, fn)
            return registered

        hook_lists = getattr(agent, "hook_lists", None)
        if isinstance(hook_lists, dict):
            for hook_name, fn in registered.items():
                hook_lists.setdefault(hook_name, []).append(fn)
            return registered

        # Fallback: bare attributes — useful in tests.
        for hook_name, fn in registered.items():
            setattr(agent, hook_name, fn)
        return registered
